astropy/tests/helper.py

```diff
@@ -192,11 +200,82 @@ def assert_quantity_allclose(actual, desired, rtol=1.0e-7, atol=None, **kwargs):
     This is a :class:`~astropy.units.Quantity`-aware version of
     :func:`numpy.testing.assert_allclose`.
     """
-    import numpy as np
-
     from astropy.units.quantity import _unquantify_allclose_arguments
 
     __tracebackhide__ = True
     np.testing.assert_allclose(
         *_unquantify_allclose_arguments(actual, desired, rtol, atol), **kwargs
     )
+
+
+def assert_objects_equal(
+    actual: AstropyArrayLike,
+    desired: AstropyArrayLike,
+    /,
+    *,
+    attrs: list[str] | None = None,
+    compare_class: bool = True,
+) -> None:
+    """Compare objects by their shape and attributes, including those provided by .info
+
+    Parameters
+    ----------
+    actual and desired: AstropyArrayLike (positional only)
+        these objects must have attributes `shape` and `info`.
+        Typically this means astropy ndarray subclasses or
+        :class:`astropy.utils.shapes.ShapedLikeNDArray`.
+
+    attrs: list[str] (optional, keyword only)
+        a list of attributes to compare in addition to the default list:
+        ['info.name', 'info.format', 'info.unit',
+        'info.description', 'info.meta', 'info.dtype']
+
+    compare_class: bool (default: False, keyword only)
+        Whether to check that the two objects are instances of the same class
+
+    Raises
+    ------
+    AssertionError
+        If actual doesn't compare equal to desired.
+    """
+    if compare_class:
+        assert actual.__class__ is desired.__class__
+
+    assert actual.shape == desired.shape
```

Comment:
`assert_quantity_allclose()` is a `Quantity`-aware wrapper for [`assert_allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) which already checks the shape. Why is this function duplicating that check?
```suggestion
```

Reply:
Nice catch.

Reply:
Actually, I don't think it's really duplicated: the loop never calls `assert_quantity_allclose` on `actual` and `desired` themselves, only on their attributes.

Reply:
I would expect that this helper would check the values in the mixin column, not just its metadata (and the function in `test_ecsv.py` does check the values). If this is only meant to check metadata then that must be reflected in the helper's name.

Reply:
In general, it does check the values - as the relevant tests pass in the relevant `attrs`. Note that there is no programmatic way currently to get what attributes contain the arrays in a class using `ShapedLikeNDArray` (with the implementation for `Time` being the one that is not quite consistent - the arrays are actually on the `TimeFormat`, which does itself inherit from `ShapedLikeNDArray`).

p.s. I'm fine with this being another reason to move it to `io.tests` rather than make it public.

Reply:
But the `ecsv` tests currently check the column values without having to specify `attrs`, so this pull request would weaken those tests.

Reply:
> But the ecsv tests currently check the column values without having to specify attrs, so this pull request would weaken those tests.

sorry I'm only seeing this now, so maybe it's harder to hop back in, but I do not get this bit; maybe I'm missing something but I don't understand how this PR, in its current state, would weaken anything, unless you are refering to your own suggestion with regards to  https://github.com/astropy/astropy/blob/8a8ee60208c79763fbde7eb3d805b8e026c1ae5f/astropy/io/ascii/tests/test_ecsv.py#L313 ?

Reply:
Suppose
```python
>>> from astropy.table import Column
>>> a = Column([4, 5])
```
With your patch the following succeeds
```python
>>> from astropy.tests.helper import assert_mixin_columns_equal
>>> assert_mixin_columns_equal(a, a + 1)
```
But on `main`
```python
>>> from astropy.io.ascii.tests.test_ecsv import assert_objects_equal
>>> assert_objects_equal(a, a + 1, [])
Traceback (most recent call last):
  ...
AssertionError
```
There are several tests that do call `assert_objects_equal(..., ..., [])`, so using your `assert_mixin_columns_equal()` as a drop-in replacement for `assert_objects_equal()` does make the tests weaker.

Reply:
Ah ! I get it. Indeed I somehow missed that `astropy.io.ascii.tests.test_ecsv.assert_objects_equal` had another part that I didn't port. Namely
```python
    # For no attrs that means we just compare directly.
    if not attrs:
        if isinstance(actual, np.ndarray) and actual.dtype.kind == "f":
            assert quantity_allclose(actual, desired, rtol=1e-15)
        else:
            assert np.all(actual == desired)
```
The good news is that included this bit back into the unified function doesn't seem to create issues on other tests that use it, so I'll add this back now.